### ISSUE TITLE
Comparison between libui-node & libui-napi

### ISSUE THREAD

**Issue by parro-it** [parro-it/libui-napi#11]
I completed the initial implementation of windows, boxes and multiline entry.
I then created a new branch from 0_3_0 branch of libui-node, removed all the stuff except windows, boxes and multiline entry.
Here is a comparison of the two implementation to help decide if the N-API rewrite is worth the effort:

## nbind version

binary size: 627Kb
compilation time: 39s

## N-API version

binary size 56Kb
compilation time: 7s

Considering also that N-API greatly simplify publish of prebuilt binaries, and that it offer much control on memory management and type conversion, my opinion is that the conversion it's totally 
worth the effort.

**Comment by mischnic** [comment]
What about compatibility?

**Comment by parro-it** [comment]
You mean at API level? It's pretty good

You can compare https://github.com/parro-it/libui-node/blob/0_3_0/examples/example-from-napi.js and https://github.com/parro-it/libui-napi/blob/master/example/using-classes/index.js

**Comment by mischnic** [comment]
Great, but I really meant node support for NAPI.

**Comment by parro-it** [comment]
Fixed!

**Comment by parro-it** [comment]
Regarding how to organize the development, I'm in doubt if it would be better to move all the staff
I already did here to a new branch in libui-node, or otherwise keep libui-napi as a separate repo and
use libui-node only for the wrapper classes.
Any thought?

**Comment by mischnic** [comment]
> use libui-node only for the wrapper classes.

What wrapper classes?

**Comment by parro-it** [comment]
These: https://github.com/parro-it/libui-napi/tree/master/js
The native part now bind libui functions one-by-one, and these classes should provide the same API as previews release.

**Comment by mischnic** [comment]
> Any thought?

Not sure.
Everything being in one place might make development easier.

**Comment by parro-it** [comment]
> Windows doesn't like the _Atomic(bool) syntax.

By now, I removed the _Atomic altogether... I should think of a better solution...

> Linux doesn't know pthread_rwlock_t.

Are we using it?

**Comment by mischnic** [comment]
> Are we using it?

No, it's used in `uv.h` and was fixed by changing `-std=c11` to `-std=gnu11`.

**Comment by parro-it** [comment]
The Atomic usage could probably be replaced by some other thread synchronization function implemented by libuv

**Comment by mischnic** [comment]
> Now I get the reason for that change...

Took quite a while to find that on the internet.

**Comment by parro-it** [comment]
> Took quite a while to find that on the internet.

I guess, there is no much documentation or example around...

I have a collection of bookmark on some reference related to node event loop and libuv docs / example / article. How could I share them?

**Comment by mischnic** [comment]
a gist?

**Comment by parro-it** [comment]
https://gist.github.com/parro-it/842ff46a86142c30b1f94f97bb738418
The titles are not saved unfortunatly... ☹️

**Comment by mischnic** [comment]
> Are you concerned on getting support for older versions?

N-API is now available in Node 6.14.2. If we want to, these changes would be required:
- use the polyfill for async-hooks (the official creditkarma/async-hooks - mine isn't needed).
- async and await aren't supported 😞 (in the download scripts).

**Comment by parro-it** [comment]
> N-API is now available in Node 6.14.2.

Awesome! Is that a new release where they backported N-API or what?

> async and await aren't supported 😞 (in the download scripts).

We can change that stuff, not a great deal...

> (the official creditkarma/async-hooks - mine isn't needed)

If I remember correctly, using your we can support node 4?

**Comment by parro-it** [comment]
> If I remember correctly, using your we can support node 4?

Ah no, because it doesn't have N-API I guess...

**Comment by mischnic** [comment]
> Awesome! Is that a new release where they backported N-API or what?

Exactly.
> > If I remember correctly, using your we can support node 4?
>
> Ah no, because it doesn't have N-API I guess...

Exactly. 😄

**Comment by mischnic** [comment]
Why is the library file called `node_libui.node` while this will be called "libui-node"?

**Comment by parro-it** [comment]
> Why is the library file called node_libui.node while this will be called "libui-node"?

Uhm, no reason...

**Comment by mischnic** [comment]
> Maybe we should support also for 32bit versions?

Not needed for macOS. How many 32bit Windows and Linux users are there?

**Comment by mischnic** [comment]
So what's your plan? If this becomes libui-node 0.4.0, what changes would be in 0.3.0?

**Comment by parro-it** [comment]
> So what's your plan? If this becomes libui-node 0.4.0, what changes would be in 0.3.0?

I thought two option here:

* Just keep in 0.3 what we already started to work on (the two existing PRs). Move everything else in 0.4.0

* Keep in 0.3.0 everything not already solved/implemented in libui-napi

I honestly prefer the first option, I fear the second will force us to implements some pieces two times...

**Comment by mischnic** [comment]
Do you want to include tables in the next libui-node (=napi) release?

**Comment by parro-it** [comment]
> Do you want to include tables in the next libui-node (=napi) release?

Yes, that's the idea. We are not too far away anyway...

**Comment by bigopon** [comment]
@parro-it @mischnic From this conversation, I understand that if I want to experiment with libui node, I should start on libui napi? If so, is it in experiment ready state?

**Comment by parro-it** [comment]
Well, the project is almost completed, I'm finishing binding of uiTables. So yes, I guess it is in an "experiment ready state".

But what are you thinking to experiment exactly? Are you trying to write a small app?

**Comment by parro-it** [comment]
The version published on npm may be not up-to-date.
And be aware that the package name will change when we'll merge back in main libui-node repo

**Comment by bigopon** [comment]
@parro-it thanks for prompt reply.
I want to experiment integrating libui node with Aurelia version next (vNext), ideally, it should be along these lines:

```ts
// app.ts
@customElement({
  name: 'app',
  template: `<template>
    <ui-form>
      <ui-entry text.bind="firstName" />
      <ui-entry text.bind="middleName" />
      <ui-entry text.bind="lastName" />
      <ui-multiline text.bind="bio" height="50" />
    </form>
  </template>`
})
export class App {
  firstName = 'Andrea';
  middleName = '';
  lastName = 'Parodi';
  bio = '';
}
```

Also able to define your own custom element.

This is to verify our renderer of version next, which has worked well so far in my experiment with [PIXIjs](http://www.pixijs.com/), [NativeScript](https://www.nativescript.org/), and server side rendering. if you are interested, [here is the repo](https://github.com/aurelia/aurelia)

About the experiment, thanks for the heads up, I'll give it a go (just finished installing 0.2 though)

**Comment by parro-it** [comment]
Oh, nice experiment... I remember I spoke with @EisenbergEffect months ago about an Aurelia renderer, but can't recall the issue ATM...

> About the experiment, thanks for the heads up, I'll give it a go (just finished installing 0.2 though)

Anyway, libui-napi will have the same public API of previuos libui-node (plus table API that we won't implement in old versions)

**Comment by bigopon** [comment]
Just installed, beside 2 warnings, I guess everything is good. Will update with result. Thanks.

**Comment by EisenbergEffect** [comment]
@parro-it It was a while ago that we talked but we're finally at a place where we can start working with what you've built. We're deep into our vNext implementation for Aurelia and we're proving out our renderer design to make sure we can render to various interesting targets besides just HTML. @bigopon is spearheading that effort and libui is next on the list :)
